linker/Linker.Steps/SweepStep.cs

```diff
@@ -308,15 +312,70 @@ protected void SweepInterfaces (TypeDefinition type)
 			}
 		}
 
-		protected void SweepCustomAttributes (ICustomAttributeProvider provider)
+		protected void SweepCustomAttributes (TypeDefinition type)
+		{
+			var removed = SweepCustomAttributes (type as ICustomAttributeProvider);
+
+			if (ShouldSetHasSecurityToFalse (type, type, type.HasSecurity, removed))
+				type.HasSecurity = false;
+		}
+
+		protected void SweepCustomAttributes (MethodDefinition method)
+		{
+			var removed = SweepCustomAttributes (method as ICustomAttributeProvider);
+
+			if (ShouldSetHasSecurityToFalse (method, method, method.HasSecurity, removed))
+				method.HasSecurity = false;
+		}
+
+		bool ShouldSetHasSecurityToFalse (ISecurityDeclarationProvider providerAsSecurity, ICustomAttributeProvider provider, bool existingHasSecurity, IList<CustomAttribute> removedAttributes)
+		{
+			if (existingHasSecurity && removedAttributes.Count > 0 && !providerAsSecurity.HasSecurityDeclarations) {
+				// If the method or type had security before and all attributes were removed, or no remaining attributes are security attributes,
+				// then we need to set HasSecurity to false
+				if (provider.CustomAttributes.Count == 0 || provider.CustomAttributes.All (attr => !IsSecurityAttributeType (attr.AttributeType.Resolve ())))
+					return true;
+			}
+
+			return false;
+		}
+
+		static bool IsSecurityAttributeType (TypeDefinition definition)
+		{
+			if (definition == null)
+				return false;
+
+			if (definition.Namespace == "System.Security") {
```

Comment:
Should probably check assembly too? Not worth it?

Reply:
I'm not to worried about it, but I think that may actually be wrong since with netcore these types would be somewhere else.